### ISSUE TITLE
global function imports

### ISSUE THREAD

**Issue by pine3ree** [silverstripe/silverstripe-framework#8331]
# Related to all namespaced versions of SilverStripe

No issue here, just an idea from other frameworks I use and have contributed to.

Should we add php global function/constants imports in order to have the php interpreter skip searching through various levels of nested namespaces?

```php
// class imports
use My\Cool\Class;
//...
// function imports
use function array_merge;
use function is_string;
use function str_replace;
//...

kind regards

**Comment by kinglozzer** [comment]
I think we’d need to see some benchmarks on this before we could make a call. Perhaps someone could run https://github.com/FriendsOfPHP/PHP-CS-Fixer with the `native_function_invocation` option enabled to see if this makes any difference?

**Comment by robbieaverill** [comment]
Agreed. If it results in noticeable improvements then we could look at it, but otherwise doing it for native functions will be detrimental to the readability of the code with tonnes more imports. I’d much rather leave it to customised functions to be imported instead, unless benchmarking shows it’s really worth it as @kinglozzer said

**Comment by robbieaverill** [comment]
How would you feel about doing some benchmarking after automating this change @pine3ree?

**Comment by pine3ree** [comment]
Hello @robbieaverill @kinglozzer, this has already been examined and benchmarked:

https://veewee.github.io/blog/optimizing-php-performance-by-fq-function-calls/
https://github.com/Roave/FunctionFQNReplacer

This turned out not to be a micro-optimization.

In zend-expressive packages all global functions are imported, that is why I suggested it. Another way is to call them with the global ns `\str_replace()`, but imo things would start to become visually unpleasant. :-)

**Comment by kinglozzer** [comment]
@pine3ree I still think we’d need to see before/after benchmarks for SilverStripe specifically. The benchmark in the [FunctionFQNReplacer repo](https://github.com/Roave/FunctionFQNReplacer#benchmark) notes, at most, a 1.3 microsecond difference per call. In a framework like ours that loads in many PHP classes, consumes (usually) file-based caches and makes a number of database queries, it may not make any discernible difference whatsoever.

I’m all in favour of performance gains though, so let’s benchmark it and hope for a good result 😄
